File: checks/ruby/lib/section508/checks/meaningful_sequence.rb
# frozen_string_literal: true

module Section508
  module Checks
    # CSS that reorders source away from reading order.
    class MeaningfulSequence
      REORDER_CLASS = /\b(pull-up|pull-left|visually-first|css-order)\b/i
      REORDER_STYLE = /(?:^|;)\s*order\s*:\s*(?!0\s*(?:;|$))/i
      REVERSE = /flex-direction\s*:\s*(row|column)-reverse/i

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        nodes = []
        nodes << root if Dom.element?(root)
        nodes.concat(Dom.elements(root, "*"))
        nodes.each do |element|
          class_name = element["class"].to_s
          style = element["style"].to_s
          if REORDER_CLASS.match?(class_name) || REORDER_STYLE.match?(style) || REVERSE.match?(style)
            return ["CSS reorders this content away from the DOM sequence."]
          end
        end
        []
      end
    end
  end
end

File: checks/python/src/section508/checks/meaningful_sequence.py
import re

from section508.dom import attr_or, class_string, inclusive_select

REORDER_CLASS = re.compile(r"\b(pull-up|pull-left|visually-first|css-order)\b", re.I)
REORDER_STYLE = re.compile(r"(?:^|;)\s*order\s*:\s*(?!0\s*(?:;|$))", re.I)
REVERSE = re.compile(r"flex-direction\s*:\s*(row|column)-reverse", re.I)


def check(root):
    if not root:
        return []

    nodes = inclusive_select(root, "*")
    for element in nodes:
        class_name = class_string(element)
        style = attr_or(element, "style")
        if REORDER_CLASS.search(class_name) or REORDER_STYLE.search(style) or REVERSE.search(style):
            return ["CSS reorders this content away from the DOM sequence."]
    return []
